### PR TITLE
Redesign Environment API

### PR DESCRIPTION
**Pull request by cswinter**
This is a large refactor of the `Environment` API. It includes [a new tutorial describing how to implement an `Environment`](https://github.com/entity-neural-network/incubator/blob/clemens/environment-api-redesign/entity_gym/TUTORIAL.md) The main change is the way `EnityID`s are handled:
- `Observation.ids` is now a `Mapping[EntityType, Sequence[EntityID]]` rather than just a `Sequence[EntityID]` to make the correspondence between entities and IDs more explicit.
- It is not required to set `Observation.ids` for all entity types, only if they are referenced by an action.
- The `*ActionMask` classes now have two ways of specifying the actors:
  - `actor_ids: Sequence[EntityID]` now takes a list of `EntityID`s as opposed to indices as was the case for `actors` before
   alternatively, `actor_types: Sequence[EntityType]` allows specifying just the types of entities that can perform the action, without listing individual ids.
  - Example: previously, one might have specified the mask as `actors = np.array([1, 2, 3])` where 1, 2, 3 happened to be the indices of all entities of some type after joined with all other entity types according to the `obs_space` ordering. Now, one could specify this as e.g. `actor_ids = [("Robot", 0), ("Robot", 1), ("Robot", 2)]` or even just `actor_types = ["Robot"]` in the case where the action is available to all entities of the "Robot" type.
  - It is permissible to omit both `actor_ids` and `actor_types`, in which case all entities can perform the action.
  - Similarly, the `SelectEntityAction` now has `actee_ids` and `actee_types`
- There is now a clear split between `Environment`, which implements a maximally convenient interface, and `VecEnv`, which implements a maximally efficient interface. In `Environment`, actors/actees are referenced via environment-specific ids, and `VecEnv` references everything with indices. The translation happens in `ListEnv`, which converts all `EntityID`s into indices when batching observations, keeps track of the `EntityID` mapping from the last observation, and converts indices back into `EntityID`s during the `act` method. As a side benefit, this pushes the work of constructing the `EntityID` lists into the subprocesses when using parallel envs which allows for additional speedups (I've observed > 22K samples/s now).

Smaller changes:
- `Observation.entities` renamed to `Observation.features`
- `Observation.action_masks` renamed to `Observation.actions`
- `Environment._act|_reset` renamed to `Environment.act|reset`, and `Environment.act|reset` renamed to `Environment.act_filter|reset_filter`
- `*Batch` renamed to `Vec*` for consistency with `VecEnv`
- The `Action` types now have separate fields for the `actors` and `actions/actees` rather than a single list that combines both, and an `items()` method that yields an iterator over the `(actor, action)` pairs
- New type aliases for `EntityType = str` and `ActionType = str` to make the intention of the API more obvious.

### CODE REVIEW

**Review by Bam4d — CHANGES_REQUESTED** [review]
Overall the new API makes more sense. 

We split out the IDS into entity types... but why don't we just put the IDS along with the entities in the features? Having the same place for both things is more coherent and much easier to debug e.g:

```
features={"box": {"id": ..., "features": [a,b,c,d,e] ....}
```
It still looks like the ordering of the IDs in the ``ids`` property is down to the environment designer, which is what I think we were aiming to avoid?

Other notes:
I made a few minor comments here and there that I found when playing with the branch.

I changed some of the tests to check that the order of the entity IDs does not effect indexing.

**Comment by cswinter** [comment]
> We split out the IDS into entity types... but why don't we just put the IDS along with the entities in the features? Having the same place for both things is more coherent and much easier to debug e.g:

I think we could do this, I actually started out combining all of "actions", "ids", and "features" into a single `EntityObs` so we have a single `entities: Mapping[EntityType, EntityObs]`, but then changed back to the previous scheme since it made it harder to figure out all the algorithmic changes. I think it probably makes sense to still keep `actions` separate, since actions feel like a top-level object and often don't make sense to split across entity type, but joining ids and entities seems good. Maybe we could also have a constructor/function on `Environment` which you pass just the lists of ids and features, and it does the whole `np.array(...` thing and validates that the number of features match and the number of ids match the number of entities.

File: entity_gym/TUTORIAL.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+                "Mine": np.array(
+                    self.mines,
+                    dtype=np.float32,
+                ).reshape(-1, 2),
+                "Robot": np.array(
+                    self.robots,
+                    dtype=np.float32,
+                ).reshape(-1, 2),
+            },
+            ids={
```
[the remark this answers is not in the document]
Comment (cswinter) [reply entity_gym/TUTORIAL.md:141]: I think at least mypy should complain.

**Comment by cswinter** [comment]
@Bam4d made a bunch of fixes and also added a new `Observation.from_entity_obs` method that allows specifying ids and features side-by-side in the same dict.

**Comment by cswinter** [comment]
It's also completely optional to use `np.array`, so you don't need to think about the right shape for empty arrays or dtype.

**Comment by Bam4d** [comment]
Ill review this by doing the necessary Griddly changes and seeing if I run into any issues!

**Comment by Bam4d** [comment]
going to try get released today so we can build and test with poetry etc... https://github.com/Bam4d/Griddly/pull/165/files

**Review by Bam4d — APPROVED** [review]

**Comment by cswinter** [comment]
@Bam4d Thanks for the fixes!

**Pull request merged**